common/src/java/org/apache/hadoop/hive/conf/HiveConf.java

```diff
@@ -6331,29 +6333,76 @@ public ZooKeeperHiveHelper getZKConfig() {
       .trustStorePassword(trustStorePassword).build();
   }
 
-  public HiveConf() {
+  public static HiveConf create() {
+    HiveConf conf = new HiveConf();
+    if (conf.isPropertyTrackingEnabled()) {
+      return new TrackedHiveConf();
+    }
+    return conf;
+  }
+
+  public static HiveConf create(Class<?> cls) {
+    HiveConf conf = new HiveConf(cls);
+    if (conf.isPropertyTrackingEnabled()) {
+      return new TrackedHiveConf(cls);
+    }
+    return conf;
+  }
+
+  public static HiveConf create(Configuration other, Class<?> cls) {
+    if (other.getBoolean(ConfVars.HIVE_CONF_PROPERTY_TRACKING.varname, false)) {
+      return new TrackedHiveConf(other, cls);
+    }
+    return new HiveConf(other, cls);
+  }
+
+  public static HiveConf create(HiveConf other) {
+    if (other.isPropertyTrackingEnabled()) {
+      return new TrackedHiveConf(other);
+    }
+    return new HiveConf(other);
+  }
+
+  private boolean isPropertyTrackingEnabled() {
+    return getBoolVar(ConfVars.HIVE_CONF_PROPERTY_TRACKING);
+  }
+
+  /**
+   * @deprecated This method will become private eventually; Use {@link #create()} instead.
+   */
+  @Deprecated
+  protected HiveConf() {
```

Comment:
I think there are other options than making HiveConf protected - and I don't think its worth to break all 3rd party extensions for a rarely used debug feature (-1)

https://issues.apache.org/jira/browse/HIVE-26985?focusedCommentId=17683330&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17683330

Reply:
hey, thanks for the comment, actually this codepath is just a dependency (before maybe merging HIVE-26984), can we discuss the constuctor related stuff on https://github.com/apache/hive/pull/3983?

Reply:
sure; I've copied almost the same message to other places... :)